lib/GHCup/Download.hs

```diff
@@ -325,6 +327,45 @@ getDownloadInfo' t v = do
       _            -> with_distro <|> without_distro_ver <|> without_distro
     )
 
+getStackDownloadInfo :: ( MonadReader env m
+                        , HasDirs env
+                        , HasGHCupInfo env
+                        , HasLog env
+                        , HasPlatformReq env
+                        , HasSettings env
+                        , MonadCatch m
+                        , MonadFail m
+                        , MonadIO m
+                        , MonadMask m
+                        , MonadThrow m
+                        )
+                     => StackSetupURLSource
+                     -> [String]
+                     -> Tool
+                     -> GHCTargetVersion
+                     -- ^ tool version
+                     -> Excepts
+                          '[NoDownload, DownloadFailed]
+                          m
+                          DownloadInfo
+getStackDownloadInfo stackSetupSource keys@(_:_) GHC tv@(GHCTargetVersion Nothing v) =
+  case stackSetupSource of
+    StackSetupURL -> do
+      (dli :: Stack.SetupInfo) <- liftE $ reThrowAll @_ @_ @'[DownloadFailed] DownloadFailed $ getBase stackSetupURL
+      let siGHCs = Stack.siGHCs $ dli
+          ghcVersionsPerKey = (\key -> M.lookup key siGHCs) <$> (T.pack <$> keys)
+      ghcVersions <- (listToMaybe . catMaybes $ ghcVersionsPerKey) ?? (NoDownload tv GHC Nothing)
+      ghcVersion <- M.lookup v ghcVersions ?? (NoDownload tv GHC Nothing)
+      fromStackDownloadInfo (Stack.gdiDownloadInfo ghcVersion)
+ where
+  fromStackDownloadInfo :: MonadThrow m => Stack.DownloadInfo -> m DownloadInfo
+  fromStackDownloadInfo Stack.DownloadInfo{..} = do
+    url <- either (\e -> throwM $ ParseError (show e)) pure $ parseURI strictURIParserOptions . E.encodeUtf8 $ downloadInfoUrl
+    sha256 <- maybe (throwM $ DigestMissing url) (pure . E.decodeUtf8) downloadInfoSha256
+    pure $ DownloadInfo url (Just $ RegexDir "ghc-.*") sha256 Nothing Nothing
+getStackDownloadInfo _ _ t v = throwE $ NoDownload v t Nothing
```

Comment:
This is the meat. Here we download the stacks setup info, index via the OS/platform keys, descend into version and then turn it into GHCup internal `DownloadInfo` type to let the rest of the callstack do the rest.